### PR TITLE
Support mashumaro DataClassORJSONMixin

### PR DESCRIPTION
**Pull request by quinten-flwls**
## Why are the changes needed?

The mashumaro `DataClassORJSONMixin` class works like the mashumaro `DataClassJSONMixin` class. Additionally, it allows for serializing dataclasses with a wider variety of attributes types, including:
- [numpy](https://numpy.org/)
- [datetime](https://docs.python.org/3/library/datetime.html#datetime.datetime)
- [date](https://docs.python.org/3/library/datetime.html#datetime.date)
- [time](https://docs.python.org/3/library/datetime.html#datetime.time)
- [uuid.UUID](https://docs.python.org/3/library/uuid.html#uuid.UUID)

Currently, flytekit does not yet support the `DataClassJSONMixin` class and throws an error when classes derived from `DataClassORJSONMixin` are serialized.

## What changes were proposed in this pull request?

This PR introduces support for classes derived from the mashumaro `DataClassORJSONMixins` class.

## How was this patch tested?

Added DataClassORJSONMixin instances in `test_type_engine.py`

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

File: flytekit/core/type_engine.py [hunk 1]
```diff
@@ -24,6 +23,7 @@
 from google.protobuf.struct_pb2 import Struct
 from marshmallow_enum import EnumField, LoadDumpOptions
 from mashumaro.mixins.json import DataClassJSONMixin
+from mashumaro.mixins.orjson import DataClassORJSONMixin
```
Comment (thomasjpfan) [inline flytekit/core/type_engine.py:26]: I prefer not to add `orjson` as a hard dependency for `flytekit`. Can you write the PR in way where `orjson` is a soft dependency?

**Review by thomasjpfan — COMMENTED** [review]
Thank you for the PR!

For the CI to run, we need to add `orjson` as a developer dependency in https://github.com/flyteorg/flytekit/blob/master/dev-requirements.in

As for testing, can you add a test that checks the expected behavior for your NumPy use case?

**Comment by quinten-flwls** [comment]
> Thank you for the PR!
> 
> For the CI to run, we need to add `orjson` as a developer dependency in https://github.com/flyteorg/flytekit/blob/master/dev-requirements.in
> 
> As for testing, can you add a test that checks the expected behavior for your NumPy use case?

Hi Thomas, thank you for your feedback! I will incorporate it. As for dev dependendies, we need both orjson and mashumaro 3.11+. What is the best way for me to specify a minimal version in the dev dependencies?

**Comment by Fatal1ty** [comment]
👋 Hi, author of mashumaro here.

> Mashumaro 3.11 introduced the DataClassORJSONMixin class

> As for dev dependendies, we need both orjson and mashumaro>=3.11. What is the best way for me to specify a minimal version in the dev dependencies?

To avoid all possible unwanted changes made on false conclusions, I’d like to make a correction. `DataClassORJSONMixin` was added earlier in 3.1 (see [release notes](https://github.com/Fatal1ty/mashumaro/releases/tag/v3.1)). I hope it’s not that important in case of this pull request and I apologize for bothering you.

File: flytekit/core/type_engine.py (hunk 1 above)
Comment (quinten-flwls) [reply flytekit/core/type_engine.py]: Great idea, orjson is a soft dependency now

**Comment by quinten-flwls** [comment]
> 👋 Hi, author of mashumaro here.
> 
> > Mashumaro 3.11 introduced the DataClassORJSONMixin class
> 
> > As for dev dependendies, we need both orjson and mashumaro>=3.11. What is the best way for me to specify a minimal version in the dev dependencies?
> 
> To avoid all possible unwanted changes made on false conclusions, I’d like to make a correction. `DataClassORJSONMixin` was added earlier in 3.1 (see [release notes](https://github.com/Fatal1ty/mashumaro/releases/tag/v3.1)). I hope it’s not that important in case of this pull request and I apologize for bothering you.

Hi @Fatal1ty! Thank you for reaching out and reminding me that `DataClassORJSONMixin` does not require mashumaro 3.11. That is encredibly useful information indeed. I updated the PR description.

The reason why I had mashumaro 3.11 in mind is because we internally use `json_encode` and `json_decode` to process `DataClassORJSONMixin` instances. But that is completely unrelated to this PR.

**Review by thomasjpfan — COMMENTED** [review]
Thanks for the update!

> wider variety of attributes, including numpy arrays.

Does your NumPy array use case work out of the box after this PR? If so, can you add a test where a NumPy array is an attribute?

**Comment by quinten-flwls** [comment]
> Thanks for the update!
> 
> > wider variety of attributes, including numpy arrays.
> 
> Does your NumPy array use case work out of the box after this PR? If so, can you add a test where a NumPy array is an attribute?

Thank you for reviewing! Unfortunately, Numpy array attributes do not work out of the box. However, there are other attribute types that do. I added a datetime attribute as an example in the tests. I hope that suffices to demonstrate the use case. Let me know if it does not!

File: flytekit/core/type_engine.py [hunk 2]
```diff
@@ -325,6 +325,13 @@ class Test(DataClassJsonMixin):
 
     def __init__(self):
         super().__init__("Object-Dataclass-Transformer", object)
+        self.serializable_classes = [DataClassJSONMixin, DataClassJsonMixin]
```
Comment (thomasjpfan) [inline flytekit/core/type_engine.py:328]: Can we make this attribute private by adding an underscore? 
(It's "private" by convention)

```suggestion
        self._serializable_classes = [DataClassJSONMixin, DataClassJsonMixin]
```

And update it in all other places.

File: tests/flytekit/unit/core/test_type_engine.py [hunk 3]
```diff
@@ -2464,6 +2497,15 @@ class Datum(DataClassJSONMixin):
...
     assert datum_mashumaro.y.value == pv.y
 
+    lt = TypeEngine.to_literal_type(DatumMashumaroORJSON)
+    now = datetime.datetime.now()
+    datum_mashumaro_orjson = DatumMashumaroORJSON(5, Color.RED, now)
+    lv = transformer.to_literal(ctx, datum_mashumaro_orjson, DatumMashumaroORJSON, lt)
+    gt = transformer.guess_python_type(lt)
+    pv = transformer.to_python_value(ctx, lv, expected_python_type=gt)
+    assert datum_mashumaro_orjson.x == pv.x
+    assert datum_mashumaro_orjson.y.value == pv.y
```
Comment (thomasjpfan) [inline tests/flytekit/unit/core/test_type_engine.py:2507]: Can we also check the value for the new datetime attribute?

```suggestion
    assert datum_mashumaro_orjson.y.value == pv.y
    assert datum_mashumaro_orjson.z.isoformat() == pv.z
```

**Review by thomasjpfan — COMMENTED** [review]
Minor nits, but otherwise LGTM

File: flytekit/core/type_engine.py (hunk 2 above)
Comment (eapolinario) [reply flytekit/core/type_engine.py]: done.

**Review by eapolinario — APPROVED** [review]

**Comment by quinten-flwls** [comment]
Thank you for the feedback and the help!

**Pull request merged**